sdk/core/azure-core/inc/http/curl/curl.hpp

```diff
@@ -12,79 +12,504 @@
+     * @param offset the number of bytes previously read.
+     * @return the number of bytes read.
+     */
+    uint64_t ReadWithOffset(uint8_t* buffer, uint64_t bufferSize, uint64_t offset);
+  };
+
+  /**
+   * @brief Concrete implementation of an HTTP Transport that uses libcurl.
+   *
+   */
+  class CurlTransport : public HttpTransport {
+  public:
+    /**
+     * @brief Implements interface to send an HTTP Request and produce an HTTP Response
+     *
+     * @param context TBD
+     * @param request an HTTP Request to be send.
+     * @return unique ptr to an HTTP Response.
+     */
     std::unique_ptr<Response> Send(Context& context, Request& request) override;
   };
 
+  /**
+   * @brief concrete implementation of a body stream to read bytes for the HTTP body using libcurl
+   * handler.
+   */
+  class CurlBodyStream : public Azure::Core::Http::BodyStream {
+  private:
+    /**
+     * @brief length of the entire HTTP Response body.
+     *
+     */
+    uint64_t m_length;
+
+    /**
+     * @brief reference to a Curl Session with all the configuration to be used to read from wire.
+     *
+     */
+    CurlSession* m_curlSession;
+
+    /**
+     * @brief Numbers of bytes already read.
+     *
+     */
+    uint64_t m_offset;
+
+  public:
+    /**
+     * @brief Construct a new Curl Body Stream object.
+     *
+     * @param length size of the HTTP Response body.
+     * @param curlSession reference to a libcurl session that contains the libcurl handler to be
+     * used.
+     */
+    CurlBodyStream(uint64_t length, CurlSession* curlSession)
+        : m_length(length), m_curlSession(curlSession), m_offset(0)
+    {
+    }
+
+    /**
+     * @brief Gets the length of the HTTP Response body.
+     *
+     * @return uint64_t
+     */
+    uint64_t Length() { return m_length; }
+
+    /**
+     * @brief Gets the number of bytes received on count from netwok. Copies the bytes to the
+     * buffer.
+     *
+     * @param buffer ptr to a buffer where to copy bytes from network.
+     * @param count number of bytes to copy from network into buffer.
+     * @return the number of read and copied bytes from network to buffer.
+     */
+    uint64_t Read(/*Context& context, */ uint8_t* buffer, uint64_t count)
+    {
+      if (m_length == m_offset)
+      {
+        return 0;
+      }
+      // Read bytes from curl into buffer. As max as the length of Stream is allowed
+      auto readCount = this->m_curlSession->ReadWithOffset(buffer, count, m_offset);
+      m_offset += readCount;
+      return readCount;
+    }
+
+    /**
+     * @brief clean up heap. Removes the libcurl session and stream from the heap.
+     *
+     * @remark calling this method deletes the stream.
+     *
+     */
+    void Close()
+    {
+      if (this->m_curlSession != nullptr)
+      {
+        delete this->m_curlSession; // Session Destructor will cleanup libcurl handle
+      }
+      // remove stream from heap
+      delete this;
```

Comment:
I feel this is really dangerous.

https://isocpp.org/wiki/faq/freestore-mgmt#delete-this

Reply:
Is this `CurlBodyStream` only supposed to be used inside Azure Core?

Reply:
It depends on the API of a service. 
For instance, storage could have an API like `BodyStream downloadBlobClient(args...)` where the stream that gets delivered to the customer is this CurlBodyStream (if customer is using libcurl).  Customer would get back one stream that abstracts the logic to read chucks of data from the HTTP body response.

Let me know if this answer you question.
Adding @JeffreyRichter, in case he want to add something more for this

Reply:
about `delete this`,  I am happy to change it. 
Somehow, the only thing we need to make sure is to delete the stream once we know it won't be used anymore.
Otherwise it would be a leak upon program termination.
The alternative is to enhance the bodyStream with a unique_ptr.  I will ask @JeffreyRichter if that's fine

Reply:
I agree, whth Jinming, `delete this` strongly smells an antipattern - it's been discouraged in books (Herb Sutter or Scott Meyers, if I am not mistaking). https://www.drdobbs.com/database/delete-this/184410022
As always in C++, there are exceptions to every rule, but here I don't think here is the case, and unique_ptr should be used.
Otherwise, you must make sure it is impossible to create an instance of CurlBodyStream on a stack: to do that, you need to make all constructors (including copy constructor) private, and expose static `create()` method, which, again, the best candidate for the return type would be `unique_ptr<CurlBodyStream>`. That would be too verbose for this specific case, so I suggest to not do `delete this`, but everywhere where CurlBodyStream is currently created, it should be created inside unique_ptr.
There is a risk an object would be reused after th's been Closed(), so after you `delete this->m_curlSession;`, assign it to nullptr:
```
delete this->m_curlSession;
this->m_curlSession = nullptr;
```